src/graphics.c

```diff
@@ -3253,6 +3280,12 @@ void line_add_word(struct line *l, struct word *word)
 	if (!l->words) {
 		l->words = word;
 		word->prev = word;
+
+		/*	This is the first word in this line. Let's use its direction
+			for the line too. Ideally, something like fribidi should be
+			used for mixing directions however. */
+		l->direction = word->direction;
```

Comment:
Sorry, but i did not found the place where you set direction in word stucture. Where is it?

Reply:
You are right. It is inside the `word_new` function which is missing from this PR.

Reply:
Like this
```
+       #ifdef _USE_HARFBUZZ
+       int u8len = u8_strlen(str);
+       uint32_t dest = 0;
+       uint32_t first[2];
+       u8_toucs(first, 2, str, u8_seqlen(str));
+       int index = 0;
+       /*      Find the first alphanumeric utf8 character for a meaningful direction
+               or use direction of the first character.
+       */
+       for (int i=0; i < u8len; i++) {
+               dest = u8_nextchar(str, &index);
+               if (!g_unichar_isalnum(dest))
+                       {
+                               dest = 0;
+                               continue;
+                       }
+       }
+
+       dest = dest?dest:first[0];
+       /* Is this made of alphabets? */
+       w->isalpha = g_unichar_isalpha(dest);
+       switch(g_unichar_get_script(dest)) {
+               case G_UNICODE_SCRIPT_ARABIC:
+                       w->direction = g_unichar_isdigit(dest)? HB_DIRECTION_LTR: HB_DIRECTION_RTL;
+                       //w->direction = HB_DIRECTION_RTL;
+                       w->script = HB_SCRIPT_ARABIC;
+                       break;
+               default:
+                       w->direction = HB_DIRECTION_LTR;
+                       w->script = HB_SCRIPT_COMMON;
+       }
+       #endif
```